controllers/nugus_controller/nugus_controller.cpp

```diff
@@ -15,107 +15,330 @@
+        if (num_ready < 0) {
+            std::cerr << "Error: Polling of TCP connection failed: " << strerror(errno) << std::endl;
+            throw;
+        }
+        else if (num_ready > 0) {
+            // Wire format
+            // unit32_t Nn  message size in bytes. The bytes are in network byte order (big endian)
+            // uint8_t * Nn  the message
+            uint32_t Nn;
+            if (recv(tcp_fd, &Nn, sizeof(Nn), 0) != sizeof(Nn)) {
+                std::cerr << "Error: Failed to read message size from TCP connection: " << strerror(errno) << std::endl;
+                throw;
+            }
 
-                // Parse message data
-                controller::nugus::RobotControl msg;
-                if (!msg.ParseFromArray(data.data(), Nh)) {
-                    std::cerr << "Error: Failed to parse serialised message" << std::endl;
-                    continue;
+            // Covert to host endianness, which might be different to network endianness
+            uint32_t Nh = ntohl(Nn);
+
+            std::vector<uint8_t> data(Nh, 0);
+            if (recv(tcp_fd, data.data(), Nh, 0) != Nh) {
+                std::cerr << "Error: Failed to read message from TCP connection: " << strerror(errno) << std::endl;
+                throw;
+            }
+
+            // Parse message data
+            ActuatorRequests actuatorRequests;
+            if (!actuatorRequests.ParseFromArray(data.data(), Nh)) {
+                std::cerr << "Error: Failed to parse serialised message" << std::endl;
+                throw;
+            }
+            //------PARSE ACTUATOR REQUESTS MESSAGE-----------
+            // For each motor in the message, get the motor and set the values for it
+            for (int i = 0; i < actuatorRequests.motor_positions_size(); i++) {
+                const MotorPID motorPID = actuatorRequests.motor_pids(i);
+                webots::Motor* motor    = this->getMotor(motorPID.name());
+                if (motor != nullptr) {
+                    if (i < actuatorRequests.motor_positions_size()) {
+                        motor->setPosition(actuatorRequests.motor_positions(i).position());
+                    }
+                    if (i < actuatorRequests.motor_velocities_size()) {
+                        motor->setVelocity(actuatorRequests.motor_velocities(i).velocity());
+                    }
+                    if (i < actuatorRequests.motor_forces_size()) {
+                        motor->setForce(actuatorRequests.motor_forces(i).force());
+                    }
+                    if (i < actuatorRequests.motor_torques_size()) {
+                        motor->setTorque(actuatorRequests.motor_torques(i).torque());
+                    }
+                    if (i < actuatorRequests.motor_pids_size()) {
+                        motor->setControlPID(motorPID.pid().x(), motorPID.pid().y(), motorPID.pid().z());
+                    }
                 }
```

Comment:
This is not the best way to handle this. We need to re-evaluate how we are going to handle these messages as it looks like forces/torques are an alternative control method to PIDs

[See this](https://cyberbotics.com/doc/reference/motor#wb_motor_set_force)
> As an alternative to the PID-controller, the wb_motor_set_force (resp. wb_motor_set_torque) function allows the user to directly specify the amount of force (resp. torque) that must be applied by a motor.

Reply:
I'm not sure what you mean, isn't this just handled by NUbots by not sending the message if we don't want to set it? Or do you mean we should have some sort of if else thing, ie don't let us set both (I'm not sure if that's what the docs are saying, that we can't set both?).